diffsmhm/diff_stats/cpu/sigma.py

```diff
@@ -0,0 +1,298 @@
+            y_periodic = np.append(y_periodic, y[p] + boxsize)
+            z_periodic = np.append(z_periodic, z[p])
+        # bottom left
+        elif x[p] < buffer_length and y[p] > boxsize - buffer_length:
+            x_periodic = np.append(x_periodic, x[p] + boxsize)
+            y_periodic = np.append(y_periodic, y[p] - boxsize)
+            z_periodic = np.append(z_periodic, z[p])
+        # top right
+        elif x[p] > boxsize-buffer_length and y[p] < buffer_length:
+            x_periodic = np.append(x_periodic, x[p] - boxsize)
+            y_periodic = np.append(y_periodic, y[p] + boxsize)
+            z_periodic = np.append(z_periodic, z[p])
+        # bottom right
+        elif x[p] > boxsize-buffer_length and y[p] > boxsize - buffer_length:
+            x_periodic = np.append(x_periodic, x[p] - boxsize)
+            y_periodic = np.append(y_periodic, y[p] - boxsize)
+            z_periodic = np.append(z_periodic, z[p])
+
+    return x_periodic, y_periodic, z_periodic
+
+
+def sigma_serial_cpu(
+    *, xh, yh, zh, wh, wh_jac, xp, yp, zp, rpbins, boxsize
+):
+    """
+    sigma_cpu_serial(...)
+        Compute the 2D (x-y) surface density with derivatives for a periodic volume.
+
+    Parameters
+    ----------
+    xh, yh, zh, wh : array-like, shape(n_halos,)
+        The arrays of positions and weights for the halos.
+    wh_jac : array-like, shape(n_grads, n_pts)
+        The array of weight gradients for the halos.
+    xp, yp, zp : array-like, shape(n_pts,)
+        The arrays of positions for the particles.
+    rpbins : array-like, shape (n_rpbins+1,)
+        Array of bin edges in the `rp` direction. Note that this array is one
+        longer than the number of bins in the `rp` direction.
+    boxsize : float
+        The size of the total periodic volume.
+
+    Returns
+    -------
+    sigma : array-like, shape(n_rpbins,)
+        The surface density at each radial bin.
+    sigma_grad : array-like, shape(n_grads, n_rpbins)
+        The gradients of the surface density.
+    """
+
+    # ensure the smallest bin is not zero
+    assert rpbins[0] > 0
+
+    # set up sizes
+    n_grads = wh_jac.shape[0]
+
+    n_rpbins = len(rpbins) - 1
+    rads = np.array(np.pi * (np.square(rpbins[1:], dtype=np.float64) -
+                    np.square(rpbins[:-1], dtype=np.float64)), dtype=np.float64)
+
+    rpmax = rpbins[-1]
+    pimax = np.ceil(boxsize).astype(int)
+
+    n_threads = int(os.environ.get("OMP_NUM_THREADS", psutil.cpu_count(logical=False)))
+
+    # handle periodicity
+    xp_p, yp_p, zp_p = _copy_periodic_points_2D(xp, yp, zp, boxsize, rpmax)
+    n_parts = len(xp_p)
+
+    # arrays to return
+    sigma_grad = np.zeros((n_grads, n_rpbins), dtype=np.float64)
+
+    # sigma
+    res = Corrfunc.theory.DDrppi(
+        autocorr=False,
+        nthreads=n_threads,
+        pimax=pimax,
+        binfile=rpbins,
+        X1=xh, Y1=yh, Z1=zh, weights1=wh,
+        periodic=False,
```

Comment:
Is there a reason you are copying points into buffers instead of letting corrfunc use the periodic BCs?

Reply:
Yeah, using the periodicity option in DDrppi is awkward here since we're counting particles along the entire volume in the z direction

Corrfunc requires that, when using periodic BCs, the `pimax` (largest bin size) for the correlation function is strictly less than half of the volume to avoid double counting. We also want the `pimax` parameter to be an integer due to the unit binning that Corrfunc does. This means that in order to use Corrfunc's periodic CBs, `pimax` must be `boxsize/2 - 1` which led to undercounting in testing. 

According to Corrfunc docs, DDrppi only counts points such that `low bin edge <= r < high bin edge` so in principle I don't see why `pimax` must be strictly less than half the domain size, but those are the current constraints in the package

Reply:
Ahhh. We should only count out to some large maximum radius. Otherwise the global domain decomposition will effectively only be 2d. A max of 40 moc/h would be fine.